File: plugins/plugin-signal/rust/src/actions/list_contacts.rs
//! List contacts action for Signal plugin.

use crate::service::SignalService;
use crate::types::get_signal_contact_display_name;
use serde::{Deserialize, Serialize};
use std::sync::Arc;

/// Parameters for listing Signal contacts (none required)
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct ListContactsParams {
    /// Whether to include blocked contacts
    #[serde(default)]
    pub include_blocked: bool,
}

/// Contact info in result
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ContactInfo {
    pub number: String,
    pub name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub uuid: Option<String>,
}

/// Result of listing Signal contacts
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ListContactsResult {
    pub success: bool,
    pub contact_count: usize,
    pub contacts: Vec<ContactInfo>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

/// Execute the list contacts action
pub async fn execute_list_contacts(
    service: Arc<SignalService>,
    params: ListContactsParams,
) -> ListContactsResult {
    match service.get_contacts().await {
        Ok(contacts) => {
            // Filter and sort contacts
            let mut filtered: Vec<_> = contacts
                .into_iter()
                .filter(|c| params.include_blocked || !c.blocked)
                .collect();

            filtered.sort_by(|a, b| {
                get_signal_contact_display_name(a)
                    .to_lowercase()
                    .cmp(&get_signal_contact_display_name(b).to_lowercase())
            });

            let contact_infos: Vec<ContactInfo> = filtered
                .iter()
                .map(|c| ContactInfo {
                    number: c.number.clone(),
                    name: get_signal_contact_display_name(c),
                    uuid: c.uuid.clone(),
                })
                .collect();

            ListContactsResult {
                success: true,
                contact_count: contact_infos.len(),
                contacts: contact_infos,
                error: None,
            }
        }
        Err(e) => ListContactsResult {
            success: false,
            contact_count: 0,
            contacts: vec![],
            error: Some(e.to_string()),
        },
    }
}

/// Format contacts as a human-readable string
pub fn format_contacts_text(result: &ListContactsResult) -> String {
    if !result.success {
        return format!(
            "Failed to list contacts: {}",
            result.error.as_deref().unwrap_or("Unknown error")
        );
    }

    if result.contacts.is_empty() {
        return "No contacts found.".to_string();
    }

    let mut lines = vec![format!("Found {} contacts:", result.contact_count)];
    lines.push(String::new());

    for contact in &result.contacts {
        lines.push(format!("• {} ({})", contact.name, contact.number));
    }

    lines.join("\n")
}

/// Action metadata
pub const ACTION_NAME: &str = "SIGNAL_LIST_CONTACTS";
pub const ACTION_DESCRIPTION: &str = "List Signal contacts";
pub const ACTION_SIMILES: &[&str] = &[
    "LIST_SIGNAL_CONTACTS",
    "SHOW_CONTACTS",
    "GET_CONTACTS",
    "SIGNAL_CONTACTS",
];

File: plugins/plugin-signal/python/elizaos_plugin_signal/actions/list_contacts.py
"""
List contacts action for Signal plugin.
"""

from typing import Optional

from elizaos_plugin_signal.types import (
    get_signal_contact_display_name,
    SIGNAL_SERVICE_NAME,
)


async def validate(runtime, message, state: Optional[dict] = None) -> bool:
    """Validate if this action can be executed."""
    return message.content.get("source") == "signal"


async def handler(
    runtime,
    message,
    state: Optional[dict] = None,
    options: Optional[dict] = None,
    callback=None,
):
    """Handle the list contacts action."""
    signal_service = runtime.get_service(SIGNAL_SERVICE_NAME)

    if not signal_service or not signal_service.is_service_connected():
        if callback:
            await callback({"text": "Signal service is not available.", "source": "signal"})
        return {"success": False, "error": "Signal service not available"}

    contacts = await signal_service.get_contacts()

    # Filter out blocked contacts and sort by name
    active_contacts = sorted(
        [c for c in contacts if not c.blocked],
        key=lambda c: get_signal_contact_display_name(c).lower(),
    )

    # Format contact list
    contact_list = [
        f"• {get_signal_contact_display_name(c)} ({c.number})"
        for c in active_contacts
    ]

    response_text = f"Found {len(active_contacts)} contacts:\n\n" + "\n".join(contact_list)

    if callback:
        await callback({
            "text": response_text,
            "source": message.content.get("source"),
        })

    return {
        "success": True,
        "data": {
            "contact_count": len(active_contacts),
            "contacts": [
                {
                    "number": c.number,
                    "name": get_signal_contact_display_name(c),
                    "uuid": c.uuid,
                }
                for c in active_contacts
            ],
        },
    }


list_contacts_action = {
    "name": "SIGNAL_LIST_CONTACTS",
    "similes": [
        "LIST_SIGNAL_CONTACTS",
        "SHOW_CONTACTS",
        "GET_CONTACTS",
        "SIGNAL_CONTACTS",
    ],
    "description": "List Signal contacts",
    "validate": validate,
    "handler": handler,
    "examples": [
        [
            {
                "name": "{{user1}}",
                "content": {"text": "Show me my Signal contacts"},
            },
            {
                "name": "{{agent}}",
                "content": {
                    "text": "I'll list your Signal contacts.",
                    "actions": ["SIGNAL_LIST_CONTACTS"],
                },
            },
        ]
    ],
}
